### ISSUE TITLE
Undetected and chattering key presses

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/firmware#1004]
@jwr reported:

> I am pretty sure there is a rare problem with keypresses not being detected or being detected multiple times. Debouncing code might need improvement.

I've also noticed this daily, but I wasn't sure until now. Maybe the key scanner thread should be given higher priority.

**Comment by kareltucek** [comment]
Actually, I dont think key scanner is the issue. I think the problem in the new discrete event handling. First I will look into related issues that we can reproduce - that may give some insight or even solve it.

**Comment by mondalaci** [comment]
This is not an issue for me anymore.

**Comment by jwr** [comment]
Hmm. This is still very much an issue for me, but I'm using 12.0.1 (for example, typing this, I just had to delete an additional '1' that appeared at the end of that version number, and then also correct the word 'appeareed'). And I can't use anything past 12.0.1. So, it's difficult for me to confirm that the issue has been fixed.

**Comment by mondalaci** [comment]
Don't you think this issue is a duplicate of #1077?

**Comment by jwr** [comment]
No, I don't think so. I had problems described in this issue for a long time now, and I still have them with 12.0.1. These problems are an annoyance, but they don't make the keyboard unusable. Also, I never saw swapped keystrokes (not in order) with 12.0.1.

What is described in #1077 started happening (in my case) after 12.0.1, and is much more severe.

That said, perhaps the code has been reworked so much that this issue no longer applies at all, but I wouldn't know that.

**Comment by mhantsch** [comment]
I agree to what @jwr said. 12.0.1 has some (occasional) repeats, but is reasonably usable. Later versions were unusable for me; I couldn't type a single meaningful sentence.

**Comment by spirodonfl** [comment]
I also have this issue and, while it's mostly consistent on which keys it affects, it does sometimes change very rarely. For the most part, the SHIFT button is not detected when I hold it down (noticeable when I'm programming). The r key is often repeated twice. Previous firmware versions had issues with the w key being repeated (no longer seems a problem) or the d key being repeated multiple times. So, while there does seem to be an improvement, it does still exist for me.

**Comment by mhantsch** [comment]
> I also have this issue

Are you on firmware 16.1.1? That's the latest release as of this time, and I have zero issues with key chatter.

If you are experiencing this behaviour with the latest firmware, you are likely experiencing a hardware issue.

**Comment by spirodonfl** [comment]
Yep latest firmware. If it's a hardware issue then I'm out of luck as I'm probably out of warranty or the return window.

**Comment by mhantsch** [comment]
Some things to check:

Are your switches in sockets (UHK60v2, UHK80) or soldered (UHK60v1)? If in sockets, you can try removing those that play up, bending their pins slightly and putting them back in the socket.

Especially for UHK60: Are only keys on the left half affected? You might have interference on the (spiral) bridge cable between the halves. 

Generally, EMV interference can cause problems with the UHK60 (especially v1). The first sets of UHK60v1 had grounding issues, see here: https://uhk.io/repair-and-modding/esd-protection-guide

There's a full guide to troubleshooting key chatter here: https://uhk.io/repair-and-modding/troubleshooting-repeated-and-missing-keystrokes

**Comment by makenomistakes** [comment]
I'm also experiencing this on firmware 16.1.1 with my UHK60 v1 (I don't recall which version I had before,
  but I didn't have this issue previously).

  Some additional context:
  - I'm using a KVM switch, but I've also tried plugging in directly via a USB 2.0 to USB-C adapter into my
  MacBook (macOS Tahoe, M4 Max) and still get this issue occasionally.
  - I sometimes get disconnects as well.
  - I have a Magic Trackpad between my two halves, but I haven't noticed that being an issue before.

  This is seriously disruptive. Also +1 to the similarity with #1077

**Comment by kareltucek** [comment]
We are indeed having some reports about some i2c reliability issues on 16.1.x, so trying 16.0.x may be a good think if you suspect software cause. 

@makenomistakes are right half keys affected too, or only left half plus modules?
